core/src/main/scala/shapeless/typeoperators.scala

```diff
@@ -19,17 +19,24 @@ package shapeless
 import scala.language.dynamics
 import scala.language.experimental.macros
 
+import scala.reflect.ClassTag
 import scala.reflect.macros.whitebox
 import scala.util.{ Try, Success, Failure }
 
-object tag {
-  def apply[U] = new Tagger[U]
-
+private[shapeless] trait Tag {
+  type @@[+T, U] <: T with Tagged[U]
```

Comment:
I have been giving this encoding a spin, for encoding/decoding tagged types to/from json:

* with the `@@` part of the private `Tag` class, I run into a problem when using play-json's format macro: 

```No instance of play.api.libs.json.Format is available for xxx.Tag.$at$at[java.lang.String, TaggingTest.FooTag] in the implicit scope (Hint: if declared in the same file, make sure it's declared before)```

* If I move the `@@` type to  the `tag` object (just a few lines below), it does work. Check-out this scastie: https://scastie.scala-lang.org/kwark/rnWsgsTQQQ2Fmsh97NMRdQ

* circe-generic seems to work with both variants

Reply:
I introduced the trait only because Scala 2.10 doesn't support abstract types in objects / classes, but somehow it supports them when inherited.

Reply:
Yes, that's true unfortunately.